incremental-font-transfer/src/patch_group.rs

```diff
@@ -0,0 +1,1253 @@
+                    iftx: ScopedGroup::NoInvalidation(no_invalidation_iftx),
+                })
+            }
+        }
+    }
+
+    /// Attempt to apply the next patch (or patches if non-invalidating) listed in this group.
+    ///
+    /// Returns the bytes of the updated font.
+    pub fn apply_next_patches(
+        self,
+        patch_data: &mut HashMap<String, UriStatus>,
+    ) -> Result<Vec<u8>, PatchingError> {
+        if let Some(patch) = self.next_invalidating_patch() {
+            let entry = patch_data
+                .get_mut(&patch.uri)
+                .ok_or(PatchingError::MissingPatches)?;
+
+            match entry {
+                UriStatus::Pending(patch_data) => {
+                    let r = self.font.apply_table_keyed_patch(patch, patch_data)?;
+                    *entry = UriStatus::Applied;
+                    return Ok(r);
+                }
+                UriStatus::Applied => {} // previously applied uris are ignored according to the spec.
+            }
+        }
+
+        // No invalidating patches left, so apply any non invalidating ones in one pass.
+        // First check if we have all of the needed data.
+        let new_font = {
+            let mut accumulated_info: Vec<(&PatchInfo, &[u8])> = vec![];
+            for info in self.non_invalidating_patch_iter() {
+                let data = patch_data
+                    .get(&info.uri)
+                    .ok_or(PatchingError::MissingPatches)?;
+
+                match data {
+                    UriStatus::Pending(data) => accumulated_info.push((info, data)),
+                    UriStatus::Applied => {} // previously applied uris are ignored according to the spec.
+                }
+            }
+
+            if accumulated_info.is_empty() {
+                return Err(PatchingError::EmptyPatchList);
+            }
+
+            self.font
+                .apply_glyph_keyed_patches(accumulated_info.into_iter())?
+        };
+
+        for info in self.non_invalidating_patch_iter() {
+            if let Some(status) = patch_data.get_mut(&info.uri) {
+                *status = UriStatus::Applied;
+            };
+        }
+
+        Ok(new_font)
+    }
+}
+
+/// Tracks whether a URI has already been applied to a font or not.
+#[derive(PartialEq, Eq, Debug)]
+pub enum UriStatus {
+    Applied,
+    Pending(Vec<u8>),
+}
+
+/// Tracks information related to a patch necessary to apply that patch.
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) struct PatchInfo {
+    uri: String,
+    source_table: IftTableTag,
+    // TODO: details for how to mark the patch applied in the mapping table (ie. bit index to flip).
+    // TODO: Signals for heuristic patch selection:
+}
+
+impl PatchInfo {
+    pub(crate) fn tag(&self) -> &IftTableTag {
+        &self.source_table
+    }
+}
+
+impl From<PatchUri> for PatchInfo {
+    fn from(value: PatchUri) -> Self {
+        PatchInfo {
+            uri: value.uri_string(),
```

Comment:
You could consider using something other than `String` to store URI, if you'r eat all concerned about the cloning costs? A simple option would be `Rc<str>`, which is a shared reference counted `str`. If you need thread safety, you could use `Arc<str>` instead.